buildSkipBlock - a Schematic block that says, I do not care what is in the world -- Ullage for Schematics

## Describe your suggestion
buildIgnoreBlocks   and   buildIgnoreExisting  do the opposite of what I am suggesting.  They ignore blocks in the world that do not match what is in the schematic.  

I want the opposite, I want to pre-process the blocks in the schematic, interpret the blocks in the schematic.  I want to add a new concept, that a particular block in the schematic tells baritone not to change anything in the world.  I do not want that to be a fixed block like Air.   I want it to be a settable block.  When this block is found in the schematic, baritone should take no action, baritone should say, I do not care what is in the world, it is OK as it is, do not try to change the world at that point.    I call this the I-do-not-care variable.   Usage would be:

#I-do-not-care  [{block_ID}/clear]

Usage to see what the I-do-not-care block is set to.
#I-do-not-care                                  

Usage to set the I-do-not-care block to be sand
#I-do-not-care sand              

Usage to clear the previous definition of the I-do-not-care block so Baritone returns to interpreting schematics literally.
#I-do-not-care clear                     

When a designer intends a schematic to use an I-do-not-care block they must choose the block at design time, and communicate this choice to users of the schematic.   The author of the schematic might save the name of the schematic to signal to the user what to set the I-do-not-care block should be:
     
My_favorite_schematic_IDC_Sand.schematic

There are so many blocks available in Minecraft that it is possible to choose any buildable block that does not appear in the intended build of the schematic.      So say I want sand to be my I-do-not-care block.   I can save the schematic, and use the schematic without the I-do-not-care-block set, and I see the oak planks being built -- this allows me to see what is in schematic.  Then I set 

#I-do-not-care sand 

and I build the schematic again -- this time not only does it not build the sand, it does not do anything at the locations that there were sand in the schematic.    

At run time, the user would know they are using a schematic with an IDC block, and they would have to set the I-do-not-care block variable appropriately before running the schematic.

The use I would put IDC blocks to is, when I want to build a schematic box to enclose a large structure, where part of the structure contains Air, I can fill the portion of the box that I do not really care about with IDC blocks.   It means I can create schematics that are effectively not right-prisms.   It means I can build triangles or pyramid-shaped schematics, and that part of the shape that is not used in the bounding-prism box is just filled with IDC blocks.   This would allow me to make effectively non-prismatic schematics with any current design tool for creating schematics. 

For instance, but certainly not limited to this use, when I am building highways in 2b2t, on axis-aligned highways, I make my schematics 6 blocks long in the highway direction.  An axis-aligned highway is a right-cylinder.   I only need to define its shape in a right-cross-section, which for minecraft, is a one-block-wide slice across the highway.   Unfortunately, one experiences glitches where blocks are laid and then disappear.  Baritone's behavior is to re-place blocks in the current schematic image that are no longer present, however, if one uses #buildrepeat, when the Baritone advances by the increment, when blocks from the last lay of the schematic, that are not in this lay of the schematic disappear, they stay as a defect after the passage of Baritone.   HighwayTools (the KAMI/Blue module) uses 2-block wide schematics, and under some conditions will leave glitches behind.   With a 6-block long schematic, every block is in 5 consecutive advances of the schematic...  In practice, only the blocks on the leading edge of the schematic are laid in a given cycle.  If there are no glitches, there is no need to re-place any blocks.  As glitches appear later and later, they may be replaced in one of the 5 previous lays.   This is a solved problem -- glitch-free axis-aligned highways -- just make the schematics longer than they have to be.   

This does not solve the problem for any highway at any angle where one does not want to determine what appears in certain blocks in the schematic -- if one does not have the use of IDC blocks.   Without IDC blocks, there is no way to have an entirely overlapping schematic that advances with buildrepeat 1,0,1 for instance without including a lot of other blocks that either Baritone must place sometimes or must remove sometimes.   When the schematic is limited only to blocks that must be determined, the blocks in the wing columns of the schematic only appear in one lay of the schematic as it advances along the other diagonal.   With the presence of glitches, there are a lot of blocks not proper in these wing columns after the pass of baritone.   The problem can be solved by enlarging the bounding prism to contain multiple appearances of all blocks in every column but then the wings of the prism extend into spaces we really do not care what is there.   No strict interpretation of the schematic will solve this problem.   

I could set a schematic for a large underground installation, and set only wall skins blocks and the contained Air blocks to be the desired block in the schematic... all the space between rooms and between tiers could be made IDC blocks, and then, if I base my schematic on a save from a world, and try to build it again at a different part of the world, Baritone is not trying to rebuild the random mineral deposits that might have appeared between rooms or between tiers in the original schematic.   Anyone who has tried to make a schematic from an existing underground build has encountered this problem.   One would be able to make standardized secret entrances (useful in Vanilla minecraft for sure) where all elements needed in some odd geometry are specified, but all unimportant parts of the bounding box are IDC blocks.   

Freeing Schematics from the Strictly-Duplicated Right-Prism Prison is what IDC blocks are about.   
      
As a further extension, I propose a second variable, As-long-as-it-is -- the full sentence would be I-do-not-care As-long-as-it-is (one of a list):

Usage:
#As-long-as-it-is [{list of block_IDs}/clear]

Usage to list the list of blocks that Baritone will apply IDC filtering to:
#As-long-as-it-is                   

Usage to set the list of blocks that Baritone will apply IDC filtering to:
#As-long-as-it-is Air Obsidian   
If the IDC block maps to a place in the world that is Air or Obsidian, do nothing.

Usage to clear the list of blocks that Baritone will apply IDC filtering to:
#As-long-as-it-is clear              

The behaviour would be, if 

#I-do-not-care sand
#As-long-as-it-is Air Obsidian

If the place in the world that sand maps to is Air or Obsidian do nothing.
In the case that the location in the world is neither Air nor Obsidian  the IDC algorithm *does* care and makes the block the first element of the #As-long-as-it-is list, in this case Air.   

I would use this particular feature for, but not limited to, when I want to dig an area in preparation to laying Obsidian.   If there is Obsidian in the schematic, and I do not have any, Baritone will fail.   If there is no Obsidian in the schematic and Obsidian is encountered in the world, Baritone will spend a lot of time mining it.   

With IDC variable, set to sand, I can just put sand in the schematic where I might eventually want to put Obsidian, and then say #As-long-as-it-is Air Obsidian -- Baritone would clear away any blocks which are not Obsidian.  I could use the build schematic, but set the IDC variable to Obsidian and the #As-long-as-it-is Air Obsidian, and get the same effect.   


Therefore:

#As-long-as-it-is Air Obsidian

is different from:

#As-long-as-it-is Obsidian Air

In the first case, the world gets more air for non-matching blocks, in the second more obsidian.  

This cannot be done by buildIgnoreBlocks   and   buildIgnoreExisting because those interpret the schematic literally and then perform judgements on the world, and only with Air either in the schematic or in the world: 


>     buildIgnoreBlocks
>     public final Settings.Setting<java.util.List<net.minecraft.block.Block>> buildIgnoreBlocks
>     A list of blocks to be treated as if they're air.
>     If a schematic asks for air at a certain position, and that position currently contains a block on this list, it will be treated as correct.
> 

>     buildIgnoreExisting
>     public final Settings.Setting<java.lang.Boolean> buildIgnoreExisting
>     If this is true, the builder will treat all non-air blocks as correct. It will only place new blocks.
> 

To do what I want to do, one has to interpret the schematic with a filter, and apply it to the world, regardless of what is in the world at that location, or only if the world does not contain certain blocks.

I got an error in the build, tried twice, decided to flush my directory and start over, still getting an error:

$ git clone https://github.com/ZacSharp/baritone.git

Cloning into 'baritone'...
remote: Enumerating objects: 35755, done.
remote: Total 35755 (delta 0), reused 0 (delta 0), pack-reused 35755
Receiving objects: 100% (35755/35755), 9.37 MiB | 1.74 MiB/s, done.
Resolving deltas: 100% (19548/19548), done.

$ cd baritone

$ git checkout horatioTesting

Branch 'horatioTesting' set up to track remote branch 'horatioTesting' from 'origin'.
Switched to a new branch 'horatioTesting'

$ ./gradlew build

> Task :buildSrc:compileJava
Note: /home/username/Desktop/Minecraft/Baritone/baritone/buildSrc/src/main/java/baritone/gradle/task/ProguardTask.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'baritone'.
> Could not resolve all artifacts for configuration ':classpath'.
   > Could not resolve org.spongepowered:mixingradle:0.6-SNAPSHOT.
     Required by:
         project :
      > Could not resolve org.spongepowered:mixingradle:0.6-SNAPSHOT.
         > Unable to load Maven meta-data from http://repo.spongepowered.org/maven/org/spongepowered/mixingradle/0.6-SNAPSHOT/maven-metadata.xml.
            > Could not HEAD 'http://repo.spongepowered.org/maven/org/spongepowered/mixingradle/0.6-SNAPSHOT/maven-metadata.xml'. Received status code 520 from server: Origin Error

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s

$

OK, is this me, or has some external source changed since I last tried to rebuild this ?

--- a/src/api/java/baritone/api/Settings.java
+++ b/src/api/java/baritone/api/Settings.java
@@ -216,6 +216,13 @@ public final class Settings {
     )));
 
     /**
+     * A mapping of blocks to blocks treated as correct in their position
+     * <p>
+     * If a schematic asks for a block on this mapping, all blocks on the mapped list will be accepted at that location as well
+     */
+    public final Setting<Map<Block, List<Block>>> buildValidSubstitutes = new Setting<>(new HashMap<>());
+
+    /**
      * A list of blocks to become air
      * <p>
      * If a schematic asks for a block on this list, only air will be accepted at that location (and nothing on buildIgnoreBlocks)

--- a/src/api/java/baritone/api/Settings.java
+++ b/src/api/java/baritone/api/Settings.java
@@ -223,6 +223,13 @@ public final class Settings {
     public final Setting<Map<Block, List<Block>>> buildValidSubstitutes = new Setting<>(new HashMap<>());
 
     /**
+     * A mapping of blocks to blocks to be built instead
+     * <p>
+     * If a schematic asks for a block on this mapping, Baritone will place the first placeable block in the mapped list
+     */
+    public final Setting<Map<Block, List<Block>>> buildSubstitutes = new Setting<>(new HashMap<>());
+
+    /**
      * A list of blocks to become air
      * <p>
      * If a schematic asks for a block on this list, only air will be accepted at that location (and nothing on buildIgnoreBlocks)

--- a/src/api/java/baritone/api/Settings.java
+++ b/src/api/java/baritone/api/Settings.java
@@ -207,6 +207,15 @@ public final class Settings {
     )));
 
     /**
+     * A list of blocks to be treated as correct.
+     * <p>
+     * If a schematic asks for any block on this list at a certain position, it will be treated as correct, regardless of what it currently is.
+     */
+    public final Setting<List<Block>> buildSkipBlocks = new Setting<>(new ArrayList<>(Arrays.asList(
+
+    )));
+
+    /**
      * A list of blocks to become air
      * <p>
      * If a schematic asks for a block on this list, only air will be accepted at that location (and nothing on buildIgnoreBlocks)

--- a/src/api/java/baritone/api/schematic/FillSchematic.java
+++ b/src/api/java/baritone/api/schematic/FillSchematic.java
@@ -19,7 +19,6 @@ package baritone.api.schematic;
 
 import baritone.api.utils.BlockOptionalMeta;
 import net.minecraft.block.state.IBlockState;
-import net.minecraft.init.Blocks;
 
 import java.util.List;
 

--- a/src/api/java/baritone/api/schematic/FillSchematic.java
+++ b/src/api/java/baritone/api/schematic/FillSchematic.java
@@ -44,8 +44,6 @@ public class FillSchematic extends AbstractSchematic {
     public IBlockState desiredState(int x, int y, int z, IBlockState current, List<IBlockState> approxPlaceable) {
         if (bom.matches(current)) {
             return current;
-        } else if (current.getBlock() != Blocks.AIR) {
-            return Blocks.AIR.getDefaultState();
         }
         for (IBlockState placeable : approxPlaceable) {
             if (bom.matches(placeable)) {

--- a/src/api/java/baritone/api/schematic/SubstituteSchematic.java
+++ b/src/api/java/baritone/api/schematic/SubstituteSchematic.java
@@ -17,7 +17,6 @@
 
 package baritone.api.schematic;
 
-import baritone.api.utils.BlockOptionalMetaLookup;
 import net.minecraft.block.Block;
 import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.state.IBlockState;

--- a/src/api/java/baritone/api/schematic/SubstituteSchematic.java
+++ b/src/api/java/baritone/api/schematic/SubstituteSchematic.java
@@ -19,8 +19,10 @@ package baritone.api.schematic;
 
 import baritone.api.utils.BlockOptionalMetaLookup;
 import net.minecraft.block.Block;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 
@@ -49,7 +51,7 @@ public class SubstituteSchematic extends AbstractSchematic {
         }
         List<Block> substitutes = substitutions.get(desiredBlock);
         if (substitutes.contains(current.getBlock()) && !current.getBlock().equals(Blocks.AIR)) {// don't preserve air, it's almost always there and almost never wanted
-            return current;
+            return withBlock(desired, current.getBlock());
         }
         for (Block substitute : substitutes) {
             if (substitute.equals(Blocks.AIR)) {
@@ -57,10 +59,25 @@ public class SubstituteSchematic extends AbstractSchematic {
             }
             for (IBlockState placeable : approxPlaceable) {
                 if (substitute.equals(placeable.getBlock())) {
-                    return placeable;
+                    return withBlock(desired, placeable.getBlock());
                 }
             }
         }
         return substitutes.get(0).getDefaultState();
+    }
+
+    private IBlockState withBlock(IBlockState state, Block block) {
+        Collection<IProperty<?>> properties = state.getPropertyKeys();
+        IBlockState newState = block.getDefaultState();
+        for (IProperty<?> property : properties) {
+            try {
+                newState = copySingleProp(state, newState, property);
+            } catch (IllegalArgumentException e) { //property does not exist for target block
+            }
+        }
+        return newState;
+    }
+    private <T extends Comparable<T>> IBlockState copySingleProp(IBlockState fromState, IBlockState toState, IProperty<T> prop) {
+        return toState.withProperty(prop, fromState.getValue(prop));
     }
 }

--- a/src/api/java/baritone/api/schematic/SubstituteSchematic.java
+++ b/src/api/java/baritone/api/schematic/SubstituteSchematic.java
@@ -23,6 +23,7 @@ import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
 import java.util.Collection;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
@@ -30,6 +31,7 @@ public class SubstituteSchematic extends AbstractSchematic {
 
     private final ISchematic schematic;
     private final Map<Block, List<Block>> substitutions;
+    private final Map<IBlockState, Map<Block, IBlockState>> blockStateCache = new HashMap<>();
 
     public SubstituteSchematic(ISchematic schematic, Map<Block,List<Block>> substitutions) {
         super(schematic.widthX(), schematic.heightY(), schematic.lengthZ());
@@ -67,6 +69,9 @@ public class SubstituteSchematic extends AbstractSchematic {
     }
 
     private IBlockState withBlock(IBlockState state, Block block) {
+        if (blockStateCache.containsKey(state) && blockStateCache.get(state).containsKey(block)) {
+            return blockStateCache.get(state).get(block);
+        }
         Collection<IProperty<?>> properties = state.getPropertyKeys();
         IBlockState newState = block.getDefaultState();
         for (IProperty<?> property : properties) {
@@ -75,6 +80,7 @@ public class SubstituteSchematic extends AbstractSchematic {
             } catch (IllegalArgumentException e) { //property does not exist for target block
             }
         }
+        blockStateCache.computeIfAbsent(state, s -> new HashMap<Block,IBlockState>()).put(block, newState);
         return newState;
     }
     private <T extends Comparable<T>> IBlockState copySingleProp(IBlockState fromState, IBlockState toState, IProperty<T> prop) {

--- a/src/api/java/baritone/api/schematic/SubstituteSchematic.java
+++ b/src/api/java/baritone/api/schematic/SubstituteSchematic.java
@@ -18,6 +18,7 @@
 package baritone.api.schematic;
 
 import net.minecraft.block.Block;
+import net.minecraft.block.BlockAir;
 import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.init.Blocks;
@@ -49,12 +50,12 @@ public class SubstituteSchematic extends AbstractSchematic {
             return desired;
         }
         List<Block> substitutes = substitutions.get(desiredBlock);
-        if (substitutes.contains(current.getBlock()) && !current.getBlock().equals(Blocks.AIR)) {// don't preserve air, it's almost always there and almost never wanted
+        if (substitutes.contains(current.getBlock()) && !(current.getBlock() instanceof BlockAir)) {// don't preserve air, it's almost always there and almost never wanted
             return withBlock(desired, current.getBlock());
         }
         for (Block substitute : substitutes) {
-            if (substitute.equals(Blocks.AIR)) {
-                return Blocks.AIR.getDefaultState(); // can always "place" air
+            if (substitute instanceof BlockAir) {
+                return current.getBlock() instanceof BlockAir ? current : Blocks.AIR.getDefaultState(); // can always "place" air
             }
             for (IBlockState placeable : approxPlaceable) {
                 if (substitute.equals(placeable.getBlock())) {

--- a/src/api/java/baritone/api/schematic/SubstituteSchematic.java
+++ b/src/api/java/baritone/api/schematic/SubstituteSchematic.java
@@ -49,23 +49,18 @@ public class SubstituteSchematic extends AbstractSchematic {
         }
         List<Block> substitutes = substitutions.get(desiredBlock);
         if (substitutes.contains(current.getBlock()) && !current.getBlock().equals(Blocks.AIR)) {// don't preserve air, it's almost always there and almost never wanted
-            System.out.println(String.format("%s is already placed", current));
             return current;
         }
         for (Block substitute : substitutes) {
             if (substitute.equals(Blocks.AIR)) {
-                System.out.println("air, lol");
                 return Blocks.AIR.getDefaultState(); // can always "place" air
             }
             for (IBlockState placeable : approxPlaceable) {
                 if (substitute.equals(placeable.getBlock())) {
-                    System.out.println(String.format("%s can be placed", placeable));
                     return placeable;
                 }
             }
-            System.out.println(String.format("%s is not an option", substitute));
         }
-        System.out.println(String.format("%s fail", substitutes.get(0).getDefaultState()));
         return substitutes.get(0).getDefaultState();
     }
 }

--- a/src/api/java/baritone/api/schematic/SubstituteSchematic.java
+++ b/src/api/java/baritone/api/schematic/SubstituteSchematic.java
@@ -1,0 +1,71 @@
+/*
+ * This file is part of Baritone.
+ *
+ * Baritone is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Baritone is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Baritone.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package baritone.api.schematic;
+
+import baritone.api.utils.BlockOptionalMetaLookup;
+import net.minecraft.block.Block;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.init.Blocks;
+import java.util.List;
+import java.util.Map;
+
+public class SubstituteSchematic extends AbstractSchematic {
+
+    private final ISchematic schematic;
+    private final Map<Block, List<Block>> substitutions;
+
+    public SubstituteSchematic(ISchematic schematic, Map<Block,List<Block>> substitutions) {
+        super(schematic.widthX(), schematic.heightY(), schematic.lengthZ());
+        this.schematic = schematic;
+        this.substitutions = substitutions;
+    }
+
+    @Override
+    public boolean inSchematic(int x, int y, int z, IBlockState currentState) {
+        return schematic.inSchematic(x, y, z, currentState);
+    }
+
+    @Override
+    public IBlockState desiredState(int x, int y, int z, IBlockState current, List<IBlockState> approxPlaceable) {
+        IBlockState desired = schematic.desiredState(x, y, z, current, approxPlaceable);
+        Block desiredBlock = desired.getBlock();
+        if (!substitutions.containsKey(desiredBlock)) {
+            return desired;
+        }
+        List<Block> substitutes = substitutions.get(desiredBlock);
+        if (substitutes.contains(current.getBlock()) && !current.getBlock().equals(Blocks.AIR)) {// don't preserve air, it's almost always there and almost never wanted
+            System.out.println(String.format("%s is already placed", current));
+            return current;
+        }
+        for (Block substitute : substitutes) {
+            if (substitute.equals(Blocks.AIR)) {
+                System.out.println("air, lol");
+                return Blocks.AIR.getDefaultState(); // can always "place" air
+            }
+            for (IBlockState placeable : approxPlaceable) {
+                if (substitute.equals(placeable.getBlock())) {
+                    System.out.println(String.format("%s can be placed", placeable));
+                    return placeable;
+                }
+            }
+            System.out.println(String.format("%s is not an option", substitute));
+        }
+        System.out.println(String.format("%s fail", substitutes.get(0).getDefaultState()));
+        return substitutes.get(0).getDefaultState();
+    }
+}

--- a/src/api/java/baritone/api/utils/SettingsUtil.java
+++ b/src/api/java/baritone/api/utils/SettingsUtil.java
@@ -35,6 +35,7 @@ import java.nio.file.NoSuchFileException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -260,6 +261,36 @@ public class SettingsUtil {
             @Override
             public boolean accepts(Type type) {
                 return List.class.isAssignableFrom(TypeUtils.resolveBaseClass(type));
+            }
+        },
+        MAPPING() {
+            @Override
+            public Object parse(ParserContext context, String raw) {
+                Type keyType = ((ParameterizedType) context.getSetting().getType()).getActualTypeArguments()[0];
+                Type valueType = ((ParameterizedType) context.getSetting().getType()).getActualTypeArguments()[1];
+                Parser keyParser = Parser.getParser(keyType);
+                Parser valueParser = Parser.getParser(valueType);
+
+                return Stream.of(raw.split(",(?=[^,]*->)"))
+                        .map(s -> s.split("->"))
+                        .collect(Collectors.toMap(s -> keyParser.parse(context, s[0]), s -> valueParser.parse(context, s[1])));
+            }
+
+            @Override
+            public String toString(ParserContext context, Object value) {
+                Type keyType = ((ParameterizedType) context.getSetting().getType()).getActualTypeArguments()[0];
+                Type valueType = ((ParameterizedType) context.getSetting().getType()).getActualTypeArguments()[1];
+                Parser keyParser = Parser.getParser(keyType);
+                Parser valueParser = Parser.getParser(valueType);
+
+                return ((Map<?,?>) value).entrySet().stream()
+                        .map(o -> keyParser.toString(context, o.getKey()) + "->" + valueParser.toString(context, o.getValue()))
+                        .collect(Collectors.joining(","));
+            }
+
+            @Override
+            public boolean accepts(Type type) {
+                return Map.class.isAssignableFrom(TypeUtils.resolveBaseClass(type));
             }
         };
 

--- a/src/main/java/baritone/process/BuilderProcess.java
+++ b/src/main/java/baritone/process/BuilderProcess.java
@@ -26,6 +26,7 @@ import baritone.api.process.IBuilderProcess;
 import baritone.api.process.PathingCommand;
 import baritone.api.process.PathingCommandType;
 import baritone.api.schematic.FillSchematic;
+import baritone.api.schematic.SubstituteSchematic;
 import baritone.api.schematic.ISchematic;
 import baritone.api.schematic.IStaticSchematic;
 import baritone.api.schematic.format.ISchematicFormat;
@@ -84,6 +85,9 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
         this.name = name;
         this.schematic = schematic;
         this.realSchematic = null;
+        if (!Baritone.settings().buildSubstitutes.value.isEmpty()) {
+            this.schematic = new SubstituteSchematic(this.schematic, Baritone.settings().buildSubstitutes.value);
+        }
         int x = origin.getX();
         int y = origin.getY();
         int z = origin.getZ();

--- a/src/main/java/baritone/process/BuilderProcess.java
+++ b/src/main/java/baritone/process/BuilderProcess.java
@@ -797,6 +797,9 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
         if (Baritone.settings().buildSkipBlocks.value.contains(desired.getBlock()) && !itemVerify) {
             return true;
         }
+        if (Baritone.settings().buildValidSubstitutes.value.getOrDefault(desired.getBlock(), Arrays.asList()).contains(current.getBlock()) && !itemVerify) {
+            return true;
+        }
         return current.equals(desired);
     }
 

--- a/src/main/java/baritone/process/BuilderProcess.java
+++ b/src/main/java/baritone/process/BuilderProcess.java
@@ -802,7 +802,7 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
         if (Baritone.settings().buildSkipBlocks.value.contains(desired.getBlock()) && !itemVerify) {
             return true;
         }
-        if (Baritone.settings().buildValidSubstitutes.value.getOrDefault(desired.getBlock(), Arrays.asList()).contains(current.getBlock()) && !itemVerify) {
+        if (Baritone.settings().buildValidSubstitutes.value.getOrDefault(desired.getBlock(), Collections.emptyList()).contains(current.getBlock()) && !itemVerify) {
             return true;
         }
         return current.equals(desired);

--- a/src/main/java/baritone/process/BuilderProcess.java
+++ b/src/main/java/baritone/process/BuilderProcess.java
@@ -794,7 +794,7 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
         if (!(current.getBlock() instanceof BlockAir) && Baritone.settings().buildIgnoreExisting.value && !itemVerify) {
             return true;
         }
-        if (Baritone.settings().buildSkipBlocks.value.contains(desired.getBlock())) {
+        if (Baritone.settings().buildSkipBlocks.value.contains(desired.getBlock()) && !itemVerify) {
             return true;
         }
         return current.equals(desired);
@@ -834,7 +834,7 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
                 return COST_INF;
             }
             IBlockState sch = getSchematic(x, y, z, current);
-            if (sch != null) {
+            if (sch != null && !Baritone.settings().buildSkipBlocks.value.contains(sch.getBlock())) {
                 // TODO this can return true even when allowPlace is off.... is that an issue?
                 if (sch.getBlock() == Blocks.AIR) {
                     // we want this to be air, but they're asking if they can place here
@@ -868,7 +868,7 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
                 return COST_INF;
             }
             IBlockState sch = getSchematic(x, y, z, current);
-            if (sch != null) {
+            if (sch != null && !Baritone.settings().buildSkipBlocks.value.contains(sch.getBlock())) {
                 if (sch.getBlock() == Blocks.AIR) {
                     // it should be air
                     // regardless of current contents, we can break it

--- a/src/main/java/baritone/process/BuilderProcess.java
+++ b/src/main/java/baritone/process/BuilderProcess.java
@@ -576,7 +576,8 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
                         continue;
                     }
                     // this is not in render distance
-                    if (!observedCompleted.contains(BetterBlockPos.longHash(blockX, blockY, blockZ))) {
+                    if (!observedCompleted.contains(BetterBlockPos.longHash(blockX, blockY, blockZ))
+                          && !Baritone.settings().buildSkipBlocks.value.contains(schematic.desiredState(x, y, z, current, this.approxPlaceable).getBlock())) {
                         // and we've never seen this position be correct
                         // therefore mark as incorrect
                         incorrectPositions.add(new BetterBlockPos(blockX, blockY, blockZ));

--- a/src/main/java/baritone/process/BuilderProcess.java
+++ b/src/main/java/baritone/process/BuilderProcess.java
@@ -794,6 +794,9 @@ public final class BuilderProcess extends BaritoneProcessHelper implements IBuil
         if (!(current.getBlock() instanceof BlockAir) && Baritone.settings().buildIgnoreExisting.value && !itemVerify) {
             return true;
         }
+        if (Baritone.settings().buildSkipBlocks.value.contains(desired.getBlock())) {
+            return true;
+        }
         return current.equals(desired);
     }
 

